File: coauthor-for-vs-code/src/agent/modelHandlerOpenAI.ts
// Standard library imports
// (none needed)

// Third-party imports
import OpenAI from 'openai';

// Local imports - utilities
import { readFile, fileExists, writeFile } from '../utils/fileUtils';
import {
  applyReplacements,
  getReplacementsByCategory,
} from '../utils/replacementUtils';
import { applyReplacementRegex } from '../utils/replacementUtils';

// Local imports - agent components
import { AgentConfig } from './AgentConfig';
import { AgentSetting, hasEndTag } from './AgentDataclass';
import { AgentStateRound } from './AgentState';
import { ModelHandler } from './ModelHandler';
import { OpenAIAPIResponseUsage, ResponseUsageFactory } from './ResponseUsage';
import { ToolState } from './ToolState';

const K_SLICE = 200;

/**
 * OpenAI-specific handlers.
 */
export class ModelHandlerOpenAI extends ModelHandler {
  /** Returns OpenAI client with configured API key. */
  getClient(): OpenAI {
    const apiKey = this.getApiKey();
    this.logger.info('Using OpenAI API.');
    return new OpenAI({ apiKey });
  }

  /** Creates a chat completion with model-specific parameters. */
  async createResponse(
    client: OpenAI,
    messages: any[],
    temperature: number,
    systemPrompt?: string,
    endTag?: string,
  ): Promise<any> {
    const kwargs: any = {
      model: this.config.fullName,
      messages,
      [this.config.name.toLowerCase().includes('o1')
        ? 'max_completion_tokens'
        : 'max_tokens']: this.config.maxOutputTokens,
    };
    if (!this.config.name.toLowerCase().includes('o1')) {
      if (endTag) {
        kwargs.stop = [endTag];
      }
      kwargs.temperature = temperature;
    }

    // Handle O1 models
    if (this.config.name.toLowerCase() === 'o1') {
      kwargs.reasoning_effort = 'high';
    }

    try {
      const response = await client.chat.completions.create(kwargs);
      return response;
    } catch (err) {
      this.logger.error(`Error in createResponse: ${err}`);
      throw err;
    }
  }

  /** Initializes message array with system prompt and user content. */
  async initializeMessages(
    userPrefix: string,
    userRequest: string,
    figureFiles?: string[],
    systemPrompt?: string,
  ): Promise<any[]> {
    const messages: any[] = [];

    // Handle system prompt differently for O1 models
    if (this.config.name.includes('o1-') || this.config.name === 'o1preview') {
      messages.push({
        role: 'user',
        content: [
          { type: 'text', text: systemPrompt },
          { type: 'text', text: userPrefix },
        ],
      });
    } else {
      if (systemPrompt) {
        // note that for openai native o1 models, they have been renamed to "developer" but "system" still works
        messages.push({ role: 'system', content: systemPrompt });
      }

      // Create content list with user prefix
      const content: any[] = [{ type: 'text', text: userPrefix }];

      // Add images if provided
      if (figureFiles) {
        content.push(...(await this.createImageMessage(figureFiles)));
      }

      messages.push({ role: 'user', content });

      // Add user request
      messages.push({
        role: this.isO1full ? 'system' : 'user',
        content: [{ type: 'text', text: userRequest }],
      });
    }

    return messages;
  }

  /** Adds user message with reflection content to existing messages. */
  createReflectionMessages(
    messages: any[],
    userMessage: string,
    figureFiles?: string[],
  ): any[] {
    const content: any[] = [];

    if (figureFiles) {
      content.push(...this.createImageContent(figureFiles));
    }
    content.push({ type: 'text', text: userMessage });
    const role = this.isO1full ? 'system' : 'user';
    messages.push({ role, content });
    return messages;
  }

  /** Formats image content for OpenAI's vision API. */
  createImageContent(imageContents: any[]): any[] {
    return imageContents.flatMap((image) => [
      { type: 'text', text: `Image: ${image.file_name}` },
      {
        type: 'image_url',
        image_url: {
          url: `data:${image.media_type};base64,${image.data}`,
          media_type: image.media_type,
          data: image.data,
        },
      },
    ]);
  }

  /** Extracts response text and usage statistics from API response. */
  extractResponse(
    responseObject: any,
    endTag: string,
    autoConfirmation = false,
  ): [string, any, string] {
    if (!responseObject.choices?.length) {
      const errorMsg = 'Invalid response from API: missing choices';
      this.logger.error(errorMsg);
      this.logger.debug(responseObject);
      throw new Error(errorMsg);
    }

    // Extract base response
    const choice = responseObject.choices[0];
    const stopReason = choice.finish_reason;
    let newResponse = choice.message.content.trim();

    // Add end tag if response was stopped and tag isn't present
    if (stopReason === 'stop' && endTag && !newResponse.includes(endTag)) {
      newResponse = `${newResponse}\n${endTag}`;
    }

    return [newResponse, responseObject.usage, stopReason];
  }

  /** Adds continuation message when response is truncated. */
  addContinueMessage(
    messages: any[],
    stateRound: AgentStateRound,
    toolState: ToolState,
    agentSetting: AgentSetting,
    agentConfig: AgentConfig,
  ): void {
    // Skip if model supports assistant prefill
    if (this.capabilities.supportsAssistantPrefill) {
      this.logger.debug(
        'Skipping continuation - assistant prefill is supported',
      );
      return;
    }

    // Create continuation message with last K tokens
    const prefillTokens = toolState.lastResponse.slice(-K_SLICE);
    const userMessageContinuation =
      `Your response got cut off, because you only have limited response space. ` +
      `Continue writing exactly from where you left off until the very end, ` +
      `marked by ${agentSetting.endTag}. ` +
      'Avoid repeat yourself and avoid starting over. ' +
      `Start your response at the next token after: "${prefillTokens}"`;

    // Add continuation message
    this.logger.info(
      'Adding continuation message to conversation. Continuation message:\n${userMessageContinuation}',
    );
    messages.push({
      role: 'user',
      content: [{ type: 'text', text: userMessageContinuation }],
    });
  }

  /** Initializes output file and handles prefill content. */
  async initializeOutputAndPrefill(
    agentConfig: AgentConfig,
    agentSetting: AgentSetting,
    messages: any[],
    toolState: ToolState,
    outputFile: string,
    prefill: string,
  ): Promise<[boolean, any[]]> {
    if (
      !(await fileExists(outputFile)) ||
      (await readFile(outputFile)).length <= 15
    ) {
      if (
        agentConfig.toolConfig.usePrefillFromInput &&
        toolState.firstKCharsFromInput
      ) {
        prefill += toolState.firstKCharsFromInput;
        toolState.updateAccumulatedOutput('');
        prefill = `<${agentSetting.documentTag}>${toolState.firstKCharsFromInput}`;
      }

      const PseudoPrefillMsgContentString = `Start your response with:\n${prefill}`;
      messages.at(-1).content.push({
        type: 'text',
        text: PseudoPrefillMsgContentString,
      });
      this.logger.info(
        `Added pseudo prefill message to messages:\n${PseudoPrefillMsgContentString}`,
      );
      return [false, messages];
    }

    let fileContent = await readFile(outputFile);
    fileContent = applyReplacements(
      fileContent,
      getReplacementsByCategory('all'),
    ).trim();
    fileContent = applyReplacementRegex(
      fileContent,
      getReplacementsByCategory('inlineMath'),
      'g',
    ).trim();

    messages.push({ role: 'assistant', content: fileContent });

    if (hasEndTag(agentSetting, fileContent)) {
      this.logger.info('End tag detected - skipping continuation');
      if (Array.isArray(messages.at(-1).content)) {
        messages.at(-1).content[messages.at(-1).content.length - 1].text =
          fileContent;
      } else {
        messages.at(-1).content = fileContent;
      }
      return [true, messages];
    }

    this.logger.info(
      'Output file exists but no end tag found - continuing from file',
    );
    if (fileContent.includes(prefill)) {
      toolState.updateAccumulatedOutput(fileContent);
    } else {
      toolState.updateAccumulatedOutput(prefill + fileContent);
      await writeFile(outputFile, toolState.accumulatedOutput);
    }
    const state = AgentStateRound.initialize(0);
    toolState.lastResponse = toolState.accumulatedOutput;
    this.addContinueMessage(
      messages,
      state,
      toolState,
      agentSetting,
      agentConfig,
    );

    return [false, messages];
  }

  /** Computes cost based on token usage and model pricing. */
  computePrice(responseUsage: any): number {
    // Handle Google models that return None for usage
    if (!responseUsage) {
      return 0.0;
    }

    // Get token counts with defaults for Google models
    const promptTokens = responseUsage.prompt_tokens ?? 0;
    const completionTokens = responseUsage.completion_tokens ?? 0;

    let basePrice =
      (promptTokens * this.config.inputPrice +
        completionTokens * this.config.outputPrice) /
      1e6;

    // Handle special token types
    if (responseUsage.reasoning_tokens) {
      basePrice +=
        (responseUsage.reasoning_tokens * this.config.outputPrice) / 1e6;
    }
    if (responseUsage.cached_tokens) {
      basePrice -=
        (responseUsage.cached_tokens * this.config.inputPrice * 0.5) / 1e6;
    }

    return basePrice;
  }

  /** Creates usage statistics from OpenAI's response format. */
  computeResponseUsage(
    responseUsage: any,
    responseTime: number,
  ): OpenAIAPIResponseUsage {
    // For Google models, create a minimal usage object with zeros
    if (!responseUsage) {
      const emptyUsage = {
        prompt_tokens: 0,
        completion_tokens: 0,
        prompt_tokens_details: { cached_tokens: 0 },
        completion_tokens_details: {
          reasoning_tokens: 0,
          accepted_prediction_tokens: null,
          rejected_prediction_tokens: null,
        },
      };
      return ResponseUsageFactory.fromOpenAIResponse(
        emptyUsage,
        this.computePrice(responseUsage),
        responseTime,
      );
    }

    return ResponseUsageFactory.fromOpenAIResponse(
      responseUsage,
      this.computePrice(responseUsage),
      responseTime,
    );
  }

  /** Updates message content with new response or continuation. */
  updateMessageContent(
    messages: any[],
    bestConnector: string,
    newResponse: string,
    toolState: ToolState,
    autoConfirmation = false,
  ): void {
    this.logger.debug(
      'Updating message content for OpenAI API compatible models',
    );

    // for OpenAI models (or models that do not support assistant prefill) the last message is always a user message
    if (messages.at(-1)?.role === 'user') {
      this.logger.debug('Last message is a user message');
      if (messages.at(-1)?.content.includes('Your response got cut off')) {
        // the second last message is an assistant message must be a assistant message
        if (messages.at(-2).role === 'assistant') {
          if (Array.isArray(messages.at(-2)?.content)) {
            messages.at(-2).content.push({
              type: 'text',
              text: bestConnector + newResponse,
            });
          } else {
            this.logger.error('Second last message content is not a list');
            messages.at(-2).content = toolState.accumulatedOutput;
          }
          // Remove continuation prompt
          messages.pop();
        }
      } else {
        this.logger.debug(
          'Last message is a request message rather than a ask to continue after cut off',
        );
        messages.push({
          role: 'assistant',
          content: [{ type: 'text', text: toolState.accumulatedOutput }],
        });
      }
    }
  }

  /** Determines if generation should continue based on response content. */
  shouldContinue(
    stopReason: string,
    newResponse: string,
    agentSetting: AgentSetting,
  ): boolean {
    // this.logger.info(
    //   'Determining if should continue for OpenAI model via OpenAI API',
    // );
    return stopReason === 'length' && !hasEndTag(agentSetting, newResponse);
  }
}

File: coauthor-python/coauthor/agent/model_handler_openai.py
"""OpenAI-specific model handlers."""

import os
from openai import OpenAI
from typing import Any

from ..logger import logger

from ..utils.file import readFile

from .agent_state import AgentStateRound
from .agent_dataclass import AgentSetting
from .agent_config import AgentConfig

from .model_handler import ModelHandler
from .response_usage import OpenAIAPIResponseUsage

from .tool_state import ToolState

K_SLICE = 200


class OpenAIHandler(ModelHandler):
    """OpenAI-specific handlers."""

    def getClient(self) -> OpenAI:
        """Get OpenAI client."""
        api_key = self.get_api_key()
        logger.info("Using OpenAI API key.")
        return OpenAI(api_key=api_key)

    def createResponse(
        self,
        client: OpenAI,
        messages: list[dict],
        temperature: float,
        systemPrompt: str | None = None,
        endTag: str | None = None,
    ) -> Any:
        """Create a response using OpenAI's API."""
        kwargs = {
            "model": self.config.fullName,
            "messages": messages,
            "max_completion_tokens" if "o1" in self.config.name.lower() else "max_tokens": self.config.maxOutputTokens,
            "temperature": 1.0 if "o1" in self.config.name.lower() else temperature,
        }

        if endTag and "o1" not in self.config.name.lower():
            kwargs["stop"] = [endTag]

        if self.config.name.lower() == "o1":
            kwargs["reasoning_effort"] = "high"

        return client.chat.completions.create(**kwargs)

    def initializeMessages(
        self,
        userPrefix: str,
        userRequest: str,
        figureFiles: list[str] | None = None,
        systemPrompt: str | None = None,
    ) -> list[dict]:
        """Initialize messages for OpenAI models."""
        messages = []

        # Handle system prompt differently for O1 models
        if self.config.name in ["o1-", "o1preview"]:
            messages = [{"role": "user", "content": [{"type": "text", "text": systemPrompt}, {"type": "text", "text": userPrefix}]}]
        else:
            if systemPrompt:
                # note that for openai native models, they have been renamed to "developer" but "system" still works
                messages.append({"role": "system", "content": systemPrompt})

            # Create content list with user prefix
            content = [{"type": "text", "text": userPrefix}]

            # Add images if provided
            if figureFiles:
                content.extend(self.create_image_message(figureFiles))

            # Add user request
            request = {"type": "text", "text": userRequest}
            content.append(request)

            messages.append({"role": "user", "content": content})

        return messages

    def create_reflection_messages(
        self,
        messages: list[dict],
        userMessage: str,
        figureFiles: list[str] | None = None,
    ) -> list[dict]:
        """Create a reflection message for OpenAI models."""
        content = []

        if figureFiles:
            content.extend(self.create_image_message(figureFiles))
        content.append({"type": "text", "text": userMessage})
        messages.append({"role": "user", "content": content})
        return messages

    def createImageContent(self, imageContents: list) -> list[dict]:
        """Create image content for OpenAI models."""

        def create_content_pair(image: dict) -> list[dict]:
            return [
                {"type": "text", "text": f"Image: {image['file_name']}"},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:{image['media_type']};base64,{image['data']}",
                        "media_type": image["media_type"],
                        "data": image["data"],
                        "detail": "high",
                    },
                },
            ]

        return [item for image in imageContents for item in create_content_pair(image)]

    def extract_response(
        self,
        responseObject: Any,
        endTag: str,
        autoConfirmation: bool = False,
    ) -> tuple[str, Any, str]:
        """Extract response text and usage statistics from OpenAI response."""
        if not (hasattr(responseObject, "choices") and responseObject.choices):
            error_msg = "Invalid response from API: missing choices"
            logger.error(error_msg)
            logger.debug(responseObject)
            raise ValueError(error_msg)

        # Extract base response
        choice = responseObject.choices[0]
        stopReason = choice.finish_reason
        newResponse = choice.message.content.strip()

        # Add end tag if response was stopped and tag isn't present
        if all([stopReason == "stop", endTag]) and endTag not in newResponse:
            newResponse = f"{newResponse}\n{endTag}"

        return newResponse, responseObject.usage, stopReason

    def add_continue_message(
        self,
        messages: list[dict],
        stateRound: AgentStateRound,
        toolState: ToolState,
        agentSetting: AgentSetting,
        agentConfig: AgentConfig,
    ) -> None:
        """Handle continuation for OpenAI models."""
        # Skip if model supports assistant prefill
        if self.capabilities.supportsAssistantPrefill:
            logger.debug("Skipping continuation - assistant prefill is supported")
            return

        # Create continuation message with last K tokens
        prefill_tokens = toolState.lastResponse[-K_SLICE:]
        userMessageContinuation = (
            f"Your response got cut off, because you only have limited response space. "
            f"Continue writing exactly from where you left off until the very end, "
            f"marked by {agentSetting.endTag}. "
            "Avoid repeat yourself and avoid starting over. "
            f'Start your response at the next token after: "{prefill_tokens}"'
        )

        # Add continuation message
        logger.info("Adding continuation message to conversation")
        logger.debug(f"Continuation message: {userMessageContinuation}")
        messages.append({"role": "user", "content": [{"type": "text", "text": userMessageContinuation}]})

    def initialize_output_and_prefill(
        self,
        agentConfig: AgentConfig,
        agentSetting: AgentSetting,
        messages: list[dict],
        toolState: ToolState,
        outputFile: str,
        prefill: str,
    ) -> tuple[bool, list[dict]]:
        """Initialize output and handle prefill for OpenAI-compatible models."""
        if not os.path.exists(outputFile) or os.path.getsize(outputFile) <= 15:
            if agentConfig.toolConfig.usePrefillFromInput and toolState.firstKCharsFromInput:
                prefill += toolState.firstKCharsFromInput
                toolState.update_accumulatedOutput("")
                prefill = f"<{agentSetting.documentTag}>{toolState.firstKCharsFromInput}"

            messages[-1]["content"].append({"type": "text", "text": f"Start your response with\n{prefill}"})
            return False, messages

        fileContent = readFile(outputFile)
        messages.append({"role": "assistant", "content": fileContent})

        if agentSetting.has_endTag(fileContent):
            logger.debug("End tag detected - skipping continuation")
            if isinstance(messages[-1]["content"], list):
                messages[-1]["content"][-1]["text"] = fileContent
            else:
                messages[-1]["content"] = fileContent
            return True, messages

        logger.warning("Output file exists but no end tag found - continuing from file")
        toolState.update_accumulatedOutput(fileContent)
        state = AgentStateRound.initialize(0)
        toolState.lastResponse = toolState.accumulatedOutput
        self.add_continue_message(messages, state, toolState, agentSetting, agentConfig)

        # here state is somehow not possible to be passed outside?
        # also here continue message is added here, not like later it was handled separately. We should make them consistent...
        return False, messages

    def computePrice(self, responseUsage: Any) -> float:
        """Compute price for OpenAI token usage."""
        # Handle Google models that return None for usage
        if responseUsage is None:
            return 0.0

        # Get token counts with defaults for Google models
        prompt_tokens = getattr(responseUsage, "prompt_tokens", 0)
        completion_tokens = getattr(responseUsage, "completion_tokens", 0)

        basePrice = (prompt_tokens * self.config.inputPrice + completion_tokens * self.config.outputPrice) / 1e6

        # Handle special token types
        if hasattr(responseUsage, "reasoning_tokens"):
            basePrice += (responseUsage.reasoning_tokens * self.config.outputPrice) / 1e6
        if hasattr(responseUsage, "cached_tokens"):
            basePrice -= (responseUsage.cached_tokens * self.config.inputPrice * 0.5) / 1e6

        return basePrice

    def computeResponseUsage(self, responseUsage: Any, responseTime: float) -> OpenAIAPIResponseUsage:
        """Compute OpenAI-specific statistics."""
        # For Google models, create a minimal usage object with zeros
        if responseUsage is None:
            return OpenAIAPIResponseUsage.from_response(
                type(
                    "EmptyUsage",
                    (),
                    {
                        "prompt_tokens": 0,
                        "completion_tokens": 0,
                        "prompt_tokens_details": type("Details", (), {"cached_tokens": 0})(),
                        "completion_tokens_details": type(
                            "Details", (), {"reasoning_tokens": 0, "accepted_prediction_tokens": None, "rejected_prediction_tokens": None}
                        )(),
                    },
                )(),
                self.computePrice(responseUsage),
                responseTime,
            )

        return OpenAIAPIResponseUsage.from_response(responseUsage, self.computePrice(responseUsage), responseTime)

    def updateMessageContent(
        self, messages: list[dict], bestConnector: str, newResponse: str, toolState: ToolState, autoConfirmation: bool = False
    ) -> None:
        """Update message content for OpenAI models."""
        logger.debug("Updating message content for OpenAI API compatible models")

        # for OpenAI models (or models that do not support assistant prefill) the last message is always a user message
        if messages[-1]["role"] == "user":
            logger.debug("Last message is a user message")
            if "Your response got cut off" in messages[-1]["content"]:
                # the second last message is an assistant message must be a assistant message
                if messages[-2]["role"] == "assistant":
                    if isinstance(messages[-2]["content"], list):
                        messages[-2]["content"].append({"type": "text", "text": bestConnector + newResponse})
                    else:
                        logger.error("Second last message content is not a list")
                        messages[-2]["content"] = toolState.accumulatedOutput
                    # Remove continuation prompt
                    messages.pop()
            else:
                logger.debug("Last message is a request message rather than a ask to continue after cut off")
                # otherwise last message is a request message rather than a ask to continue after cut off
                messages.append({"role": "assistant", "content": [{"type": "text", "text": toolState.accumulatedOutput}]})

    def should_continue(self, stopReason: str, newResponse: str, agentSetting: AgentSetting) -> bool:
        """Determine if OpenAI model should continue generating."""
        logger.info("Determining if should continue for OpenAI model via OpenAI API")
        return stopReason == "length" and not agentSetting.has_endTag(newResponse)
